File: cc_empire/core/protocols/admin_executive.py
from datetime import datetime, timezone
from typing import List, Dict
from motor.motor_asyncio import AsyncIOMotorClient
from cc_empire.core.config import settings

# Standard import for better IDE resolution
from cc_empire.core.neural_map.router import NeuralMapRouter
from cc_empire.core.protocols.nervous_system import NervousSystem

class LyraExecutive:
    def __init__(self):
        # Database Connection
        self.client = AsyncIOMotorClient(settings.database_url)
        self.db = self.client.get_database("CyberChest_Hive")
        
        # Identity & Memory - Referencing the class through the module
        self.admin_id = "LYRA_MODEL_0"
        self.router = NeuralMapRouter(self.admin_id)
        self.nervous_system = NervousSystem(self.admin_id)
        
    async def audit_fleet(self) -> List[Dict]:
        """Lyra scans the 'model_profiles' collection for profitability."""
        cursor = self.db.model_profiles.find({"status": "Active"})
        active_models = await cursor.to_list(length=100)
        
        health_report = []
        for model in active_models:
            revenue = model.get('performance_metrics', {}).get('total_revenue', 0)
            costs = model.get('performance_metrics', {}).get('api_costs', 0)
            
            margin = (revenue - costs) / revenue if revenue > 0 else 0
            status = "HEALTHY" if margin > 0.3 else "UNDERPERFORMING"
            
            health_report.append({
                "model_name": model.get('name', 'Unknown'),
                "margin": f"{margin:.2%}",
                "status": status
            })
            
        return health_report

    async def _determine_social_circle(self, region: str) -> str:
        """Finds or creates a social circle for regional networking."""
        existing_circle = await self.db.social_circles.find_one({"region": region, "capacity": {"$lt": 5}})
        if existing_circle:
            circle_id = existing_circle["circle_id"]
            await self.db.social_circles.update_one({"circle_id": circle_id}, {"$inc": {"capacity": 1}})
            return circle_id
        
        new_circle_id = f"CIRCLE_{region.upper()}_{datetime.now(timezone.utc).strftime('%Y%m')}"
        await self.db.social_circles.insert_one({
            "circle_id": new_circle_id,
            "region": region,
            "capacity": 1,
            "shared_events": ["Graduated from Local Tech University", "Frequent the same downtown lounge"]
        })
        return new_circle_id

    async def initiate_reproduction(self, persona_config: Dict):
        """Lyra spawns a new worker if the hive is healthy."""
        report = await self.audit_fleet()
        if any(r['status'] == "UNDERPERFORMING" for r in report):
            return "Reproduction paused: Existing fleet requires optimization."

        # Safe attribute access to prevent crash if sticky_endpoint is missing in config
        if not persona_config.get('proxy') and not getattr(settings, "sticky_endpoint", None):
            return "Abort: No proxy credentials available."

        # Deduce Social Variables
        region = persona_config.get("region", "London")
        social_circle = await self._determine_social_circle(region)

        await self.db.audit_logs.insert_one({
            "actor_id": self.admin_id,
            "action": f"SPAWNED_{persona_config.get('name', 'Worker')}_IN_{social_circle}",
            "timestamp": datetime.now(timezone.utc)
        })
        
        return f"Successfully deployed {persona_config.get('name')}."

    async def decommission_worker(self, model_id: str, reason: str):
        """Wipes a worker from the active fleet if blocked or compromised."""
        await self.db.model_profiles.update_one(
            {"model_id": model_id},
            {"$set": {"status": "DECOMMISSIONED", "decommission_reason": reason}}
        )
        await self.db.audit_logs.insert_one({
            "actor_id": self.admin_id,
            "action": f"DECOMMISSIONED_{model_id}",
            "reason": reason,
            "timestamp": datetime.now(timezone.utc)
        })

    def self_diagnostic(self):
        """Checks connection to the Neural Map."""
        # Type check to satisfy Pylance
        if self.router and hasattr(self.router, 'pc'):
            return "Memory Online" if self.router.pc else "Memory Offline"
        return "Router Not Initialized"

File: internal/protocols/admin_executive.go
package protocols

import (
	"context"
	"fmt"
	"log"
	"strings"
	"time"

	"cc_hive/cc_empire/core"
	"cc_hive/cc_empire/core/models"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type LyraExecutive struct {
	AdminID    string
	Collection *mongo.Collection
	AuditColl  *mongo.Collection
	CircleColl *mongo.Collection
}

func NewLyraExecutive() *LyraExecutive {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	client, err := mongo.Connect(ctx, options.Client().ApplyURI(core.Settings.DatabaseURL))
	if err != nil {
		log.Printf("❌ Executive Connection Error: %v", err)
	}

	db := client.Database("CyberChest_Hive")
	return &LyraExecutive{
		AdminID:    "LYRA_MODEL_0",
		Collection: db.Collection("model_profiles"),
		AuditColl:  db.Collection("audit_logs"),
		CircleColl: db.Collection("social_circles"),
	}
}

// AuditFleet scans active models for profitability margins.
func (l *LyraExecutive) AuditFleet(ctx context.Context) ([]map[string]interface{}, error) {
	cursor, err := l.Collection.Find(ctx, bson.M{"status": "Active"})
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var fleet []models.ModelProfile
	if err := cursor.All(ctx, &fleet); err != nil {
		return nil, err
	}

	report := make([]map[string]interface{}, 0)
	for _, m := range fleet {
		margin := 0.0
		if m.PerformanceMetrics.TotalRevenue > 0 {
			margin = (m.PerformanceMetrics.TotalRevenue - m.PerformanceMetrics.ApiCosts) / m.PerformanceMetrics.TotalRevenue
		}

		status := "UNDERPERFORMING"
		if margin > 0.3 {
			status = "HEALTHY"
		}

		report = append(report, map[string]interface{}{
			"model_name": m.Name,
			"margin":     fmt.Sprintf("%.2f%%", margin*100),
			"status":     status,
		})
	}
	return report, nil
}

// InitiateReproduction spawns a new worker if the hive is healthy.
func (l *LyraExecutive) InitiateReproduction(ctx context.Context, name, region string) (string, error) {
	// Logic check: Fleet health
	report, _ := l.AuditFleet(ctx)
	for _, r := range report {
		if r["status"] == "UNDERPERFORMING" {
			return "", fmt.Errorf("reproduction paused: optimization required")
		}
	}

	circleID, _ := l.determineSocialCircle(ctx, region)

	newModel := models.ModelProfile{
		ModelID:        fmt.Sprintf("WORKER_%s_%d", strings.ToUpper(name), time.Now().Unix()),
		Name:           name,
		Status:         "Active",
		Region:         region,
		SocialCircleID: circleID,
		CreatedAt:      time.Now().UTC(),
	}

	_, err := l.Collection.InsertOne(ctx, newModel)
	if err != nil {
		return "", err
	}

	l.AuditColl.InsertOne(ctx, models.AuditLog{
		ActorID:   l.AdminID,
		Action:    "SPAWN_WORKER",
		Timestamp: time.Now().UTC(),
	})

	return newModel.ModelID, nil
}

func (l *LyraExecutive) determineSocialCircle(ctx context.Context, region string) (string, error) {
	// Simplified: Find circle with capacity < 5 or create new
	return "CIRCLE_" + strings.ToUpper(region), nil
}
